### PR TITLE
Make Calls To Backend In Parallel

### PR DESCRIPTION
**Pull request by guscarreon**
The benchmarking tests that this PR wrote into `endpoints/integration_test.go` show less nanoseconds per operation for the parallel version of `NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool)` than its current sequential version.

**Non-parallel** `NewPutHandler`
```
╰─➤  go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-cache/endpoints
BenchmarkPutHandlerLen1-12        602461              2034 ns/op
BenchmarkPutHandlerLen2-12        579069              2199 ns/op
BenchmarkPutHandlerLen4-12        450772              2327 ns/op
BenchmarkPutHandlerLen8-12        532084              2277 ns/op
PASS
ok      github.com/prebid/prebid-cache/endpoints        63.661s
```

**Parallel** `NewPutHandler`
```
╰─➤  go test -bench=.                                                                                               2 ↵
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-cache/endpoints
BenchmarkPutHandlerLen1-12        592201        1886 ns/op
BenchmarkPutHandlerLen2-12        642481              1938 ns/op
BenchmarkPutHandlerLen4-12        606862              1950 ns/op
BenchmarkPutHandlerLen8-12        603174              2104 ns/op
PASS
ok      github.com/prebid/prebid-cache/endpoints        66.183s
```

### CODE REVIEW

File: endpoints/put.go [hunk 1]
```diff
@@ -9,128 +9,99 @@ import (
...
-			// Eventually we may want to provide error details, but as of today this is the only non-fatal error
-			// Future error details could go into a second property of the Responses object, such as "errors"
-			if len(resps.Responses[i].UUID) > 0 {
-				err = backend.Put(ctx, resps.Responses[i].UUID, toCache, p.TTLSeconds)
-				if err != nil {
-					if _, ok := err.(*backendDecorators.BadPayloadSize); ok {
-						http.Error(w, fmt.Sprintf("POST /cache element %d exceeded max size: %v", i, err), http.StatusBadRequest)
-						return
+					// Run process in parallel
+					getIndexChannel := make(chan int)
```
Comment (SyntaxNode) [inline endpoints/put.go]: A single channel should be enough. Each go routine running in parallel needs to 1) possibly make a get call to the backend (to verify a user provided key is unique.. although that's an issue for another day) and then 2) always make a put call to the backend.

Since steps 1 and 2 always must be synchronous within each parallelized put, I don't see a reason to separate those actions and take additional overhead cost. Kicking off all go routines and waiting for them to complete should suffice.

Taking this a step further, we don't need to use channels at all. Each concurrent goroutine can store their result in the array at a predetermined index, as you have in this code, so that shouldn't cause any memory issues. You can use a wait group to coordinate completion instead, which should be a bit faster.

File: endpoints/put.go [hunk 2]
```diff
@@ -32,113 +31,199 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
...
+		w.Write(bytes)
+	}
+}
+
+func validateAndEncode(puts *PutRequest, resp *PutResponse, exit *exitInfo) {
+	for index, p := range puts.Puts {
+		var toCache string
+
+		if len(p.Value) == 0 {
+			exit.errMsg = "Missing value."
```
Comment (SyntaxNode) [inline endpoints/put.go]: Consider returning new exitInfo objects instead of alerting an existing one. You can avoid memory duplication by doing the same in the other `if` branchea.

File: endpoints/put.go [hunk 3]
```diff
@@ -32,113 +31,199 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
...
+type backendCallObject struct {
+	client    backends.Backend
+	put       *PutRequest
+	resp      *PutResponse
+	allowKeys bool
+	ctx       context.Context
+}
+
+func callBackendInParallel(backend *backendCallObject, exit *exitInfo, indexChannel <-chan int, done chan<- bool) {
+	for index := range indexChannel {
```
Comment (SyntaxNode) [inline endpoints/put.go]: I'm not sure this provides parallelization. Instead of passing the indexChannel here, consider calling this method in each iteration of the callsite.

Something like:
```
func callBackend(backend *backendCallObject, exit *exitInfo, index int, done chan<- bool) {
  defer func() { done <- true }
  callBackendGet(backend, index)
  callBackendPut(backend, exit, index)
}
```

.. and at the call site ...

```
// start workers
for i, _ := range backend.put.Puts {
   go callBackend(backend, exitData, i, done)
}

// wait for all workers to be done
for _ , _ := range backend.put.Puts {
  <-done
}
```

File: endpoints/put.go (hunk 1 above)
Comment (guscarreon) [reply endpoints/put.go]: Done. USed `WaitGroup` instead of channels and the code really looks cleaner

File: endpoints/put.go (hunk 2 above)
Comment (guscarreon) [reply endpoints/put.go:118]: `http.StatusOK` gets assigned to the `exitData` object in line 63 and we test for the status to keep being OK in lines  74 and 92.  Isn't it beter to pass the `exitData` pointer to both `validateAndEncode(backend *backendCallObject, exit *exitInfo)` and `callBackendPut(backend *backendCallObject, exit *exitInfo, index int)`? I suppose the pointer approach will create a single memory allocation in `NewPutHandler`'s stack and by passing by reference to `validateAndEncode(backend *backendCallObject, exit *exitInfo)` and `callBackendPut(backend *backendCallObject, exit *exitInfo, index int)`, aren't we only creating `int`s in their respective stacks as opposed to additional `exitInfo` objects in their respective stacks. I can change it if you want though.

**Comment by guscarreon** [comment]
Increased coverage and parallelism benchmark improved:
```
╰─➤  go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-cache/endpoints
BenchmarkPutHandlerLen1-12        482808              2336 ns/op
BenchmarkPutHandlerLen2-12        512317              2424 ns/op
BenchmarkPutHandlerLen4-12        476258              2574 ns/op
BenchmarkPutHandlerLen8-12        388086              2698 ns/op
PASS
ok      github.com/prebid/prebid-cache/endpoints        57.368s
```
**Coverage**
```
╰─➤  go test -coverprofile=c.out
ERRO[0000] POST /cache Error while writing to the backend: This is a mock backend that returns this error on Put() operation
ERRO[0000] POST /cache had an unexpected error:This is a mock backend that returns this error on Put() operation
ERRO[0000] POST /cache Error while writing to the backend: context deadline exceeded
ERRO[0000] POST /cache timed out:context deadline exceeded
PASS
coverage: 92.0% of statements
ok      github.com/prebid/prebid-cache/endpoints        0.081s

╰─➤  go tool cover -func=c.out
github.com/prebid/prebid-cache/endpoints/get.go:15:     NewGetHandler           95.0%
github.com/prebid/prebid-cache/endpoints/get.go:51:     parseUUID               100.0%
github.com/prebid/prebid-cache/endpoints/index.go:11:   Index                   0.0%
github.com/prebid/prebid-cache/endpoints/put.go:20:     NewPutHandler           92.0%
github.com/prebid/prebid-cache/endpoints/put.go:109:    validateAndEncode       90.6%
github.com/prebid/prebid-cache/endpoints/put.go:164:    callBackendInParallel   100.0%
github.com/prebid/prebid-cache/endpoints/put.go:170:    callBackendGet          80.0%
github.com/prebid/prebid-cache/endpoints/put.go:181:    callBackendPut          100.0%
github.com/prebid/prebid-cache/endpoints/status.go:8:   Status                  100.0%
total:                                                  (statements)            92.0%
```

File: endpoints/put.go [hunk 4]
```diff
@@ -32,105 +31,70 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
...
-				if err != nil || len(s) == 0 {
-					resps.Responses[i].UUID = p.Key
+		// Get a response object from the resource pool that we'll fill with processed info
+		backend.resp = putResponsePool.Get().(*PutResponse)
+		backend.resp.Responses = make([]PutResponseObject, len(backend.put.Puts))
+		backend.resp.toCacheStrings = make([]string, len(backend.put.Puts))
+		defer putResponsePool.Put(backend.resp)
+
+		var cancel func()
+		backend.ctx, cancel = context.WithTimeout(context.Background(), 500*time.Millisecond)
```
Comment (hhhjort) [inline endpoints/put.go:60]: Doesn't this create a context that expires 500 milliseconds after midnight on Jan 1, 1970?

Comment (SyntaxNode) [reply endpoints/put.go:60]: I don't see why it would. WithTimeout takes a duration, not a timestamp.

File: endpoints/put.go [hunk 5]
```diff
@@ -142,6 +106,105 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
...
+		if err != nil || len(s) == 0 {
+			backend.resp.Responses[index].UUID = backend.put.Puts[index].Key
+		} else {
+			backend.resp.Responses[index].UUID = ""
+		}
+	}
+}
+
+func callBackendPut(backend *backendCallObject, exit *exitInfo, index int) {
+	if len(backend.resp.Responses[index].UUID) > 0 {
```
Comment (SyntaxNode) [inline endpoints/put.go:182]: What's happening here is a bit hard to follow. I think you refactored the code well and this was hard to follow before as well. I think we can leave it alone for now, but consider revisiting in a future story to make it clear that this handles the case of a Key was specified but does not exist. Should be handled as an err instead of empty string.

Comment (guscarreon) [reply endpoints/put.go:182]: Totally agree. I tried to leave the logic as is and only focus in the parallelism part, but I agree that the `if err != nil || len(s) == 0` statment could be dissected to differentiate between one option or the other. Maybe, in case of an error, throw it and interrupt execution if the data service is down.

**Comment by guscarreon** [comment]
> Main flow looks good. I'll review the test cases tomorrow. Might want to run it through race detection checks a bit to make sure we did't overlook anything.
> 
> I think the exitMsg and status would 'flow' better (easier to understand function inputs and outputs) if changed to two returns instead of alerting an existing object. They should be passed via the stack and likely not have a significant run time impact. I'm suggesting to save this optimization for if performance analysis finds it necessary.

Sure, we can discuss further optimization or refactoring. I agree that it still looks like a lot and the entire thing could be redesigned to look cleaner if we wanted. On the race data department, I doubled checked and Go seems to be ok so far. Do you think I should design/perform further race condition tests? I thought the `sync.Pool`s got us covered on that front.
```
╭─ ~/go/src/github.com/prebid/prebid-cache  ‹PBS-614-C*›
╰─➤  go build -race
╭─ ~/go/src/github.com/prebid/prebid-cache  ‹PBS-614-C*›
╰─➤  ./validate.sh --nofmt --cov --race 50
?       github.com/prebid/prebid-cache  [no test files]
ok      github.com/prebid/prebid-cache/backends 0.023s
?       github.com/prebid/prebid-cache/backends/config  [no test files]
ok      github.com/prebid/prebid-cache/backends/decorators      0.026s
?       github.com/prebid/prebid-cache/compression      [no test files]
ok      github.com/prebid/prebid-cache/config   0.029s
ok      github.com/prebid/prebid-cache/endpoints        0.077s
ok      github.com/prebid/prebid-cache/endpoints/decorators     0.026s
?       github.com/prebid/prebid-cache/endpoints/routing        [no test files]
?       github.com/prebid/prebid-cache/metrics  [no test files]
?       github.com/prebid/prebid-cache/metrics/metricstest      [no test files]
ok      github.com/prebid/prebid-cache/server   0.023s
```

**Comment by SyntaxNode** [comment]
Work will continue and a new PR will be opened later.
